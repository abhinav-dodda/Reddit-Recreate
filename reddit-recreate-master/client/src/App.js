import React, { Component } from 'react'
import NewsList from './components/NewsList/NewsList'
import throttle from 'lodash.throttle'
import logo from './logo.png'; 

import './App.css'

export default class App extends Component {
  constructor() {
    super()
    this.state = {
      limit: 25,
      after_news: "",
      newsList: [],     
      loading: false,
      intervalID: 0,
      news_count: 0
    }
    this.handleScroll = throttle(this.handleScroll, 50)
  }

  componentDidMount = () => {
    document.addEventListener('scroll', this.handleScroll)
    this.fetchNews(this.state.after_news)

    let self = this;
    let timer_id = setInterval(function(){      
      self.fetchPeriod();
    }, 60 * 1000);
    console.log(timer_id);
    this.setState({intervalID: timer_id});
  }

  fetchPeriod = async() => {
    this.setState({ loading: true })
    let count = this.state.news_count;
    let api_path = "https://www.reddit.com/r/news.json?limit=" + count;
    const res = await fetch(api_path)
    const parseRes = await res.json()

    let curDate = new Date(); 
    let curTimestamp = curDate.getTime() / 1000;
   
    for(let i = 0; i < parseRes.data.children.length; i++){                   
      let created_utc = parseRes.data.children[i].data.created_utc;      
      parseRes.data.children[i].data.created_diff = this.timeDifference(curTimestamp, created_utc);  

      let url = parseRes.data.children[i].data.url;
      
      url = url.indexOf("https://www.") > -1 ? url.replace("https://www.", "") : url;
      url = url.indexOf("https://") > -1 ? url.replace("https://", "") : url;
      url = url.indexOf("http://") > -1 ? url.replace("http://", "") : url;
      url = url.indexOf("www.") > -1 ? url.replace("www.", "") : url;

      let pos = url.indexOf("/");      
      parseRes.data.children[i].data.url_short = url.substring(0, pos) + url.substring(pos, pos + 7) + "...";
    }

    this.setState(prevState => ({    
      after_news: parseRes.data.after,  
      newsList: parseRes.data.children,
      loading: false,
      news_count: count
    }))

  }

  componentWillUnmount = () => {
    document.removeEventListener('scroll', this.handleScroll)

    clearInterval(this.state.intervalID);
  }

  timeDifference = (timestamp1, timestamp2) => {
    let difference = timestamp1 - timestamp2;

    let daysDifference = Math.floor(difference/60/60/24);
    difference -= daysDifference*1000*60*60*24

    let hoursDifference = Math.floor(difference/60/60);
    difference -= hoursDifference*1000*60*60

    let minutesDifference = Math.floor(difference/60);
    difference -= minutesDifference*1000*60

    let secondsDifference = Math.floor(difference);

    if(daysDifference > 0) {    
      return daysDifference + " days ago"
    }else if(hoursDifference > 0) {
      return hoursDifference + " hours ago"
    }else if(minutesDifference > 0) {
      return minutesDifference + " minutes ago"
    }else if(secondsDifference > 0) {
      return secondsDifference + " seconds ago"
    }
     return ""
  }

  fetchNews = async after_news => {
    this.setState({ loading: true })

    let after = this.state.after_news === "" ? "" : "&after=" + this.state.after_news;
    let api_path = "https://www.reddit.com/r/news.json?limit=" + this.state.limit + after;

    const res = await fetch(api_path)
    const parseRes = await res.json()

    let curDate = new Date(); 
    let curTimestamp = curDate.getTime() / 1000;
   
    for(let i = 0; i < parseRes.data.children.length; i++){                   
      let created_utc = parseRes.data.children[i].data.created_utc;      
      parseRes.data.children[i].data.created_diff = this.timeDifference(curTimestamp, created_utc);  

      let url = parseRes.data.children[i].data.url;
      
      url = url.indexOf("https://www.") > -1 ? url.replace("https://www.", "") : url;
      url = url.indexOf("https://") > -1 ? url.replace("https://", "") : url;
      url = url.indexOf("http://") > -1 ? url.replace("http://", "") : url;
      url = url.indexOf("www.") > -1 ? url.replace("www.", "") : url;

      let pos = url.indexOf("/");      
      parseRes.data.children[i].data.url_short = url.substring(0, pos) + url.substring(pos, pos + 7) + "...";
    }
    let count = this.state.news_count + 25;
    this.setState(prevState => ({
      after_news: parseRes.data.after,
      newsList: [...prevState.newsList, ...parseRes.data.children],
      loading: false,
      news_count: count
    }))

  }

  handleScroll = () => {
    if (window.innerHeight + window.pageYOffset >= document.body.offsetHeight - 550) {
      if (!this.state.loading) {
        this.fetchNews(this.state.after_news)
      }
    }
  }

  render() {
    return (
      <div className="App">       
        <div className="header">
          <img src={logo} alt="logo" />
        </div>
        <div className="container">
          <NewsList {...this.state} />
        </div>
      </div>
    )
  }
}
