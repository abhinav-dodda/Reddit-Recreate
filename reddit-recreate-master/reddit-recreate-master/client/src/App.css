* {
  margin: 0;
  padding: 0;
  -webkit-box-sizing: border-box;
  box-sizing: border-box;
}

body {
  font-family: 'Quicksand', sans-serif;
  color: #7e7e7e;
}

.App {
  background-color: #DAE0E6;
}

.header {
  height: 10vh;
  background-color: #42ADF0;
  margin-bottom: 20px;
}

.header img{
  height: 3rem;
  width: 9rem;
  margin-left: 1rem;
  margin-top: 0.7rem;
}

.container {
  max-width: 900px;
  margin: 0 auto;
  padding: 0 1rem;
}

.code {
  cursor: pointer;
  position: fixed;
  top: 10%;
  right: 30px;
  z-index: 5;
}

.code img {
  width: 48px;
}

.shadow {
  -webkit-box-shadow: 0px 4px 6px 0 hsla(0,0%,0%,0.2);
  box-shadow: 0px 4px 6px 0 hsla(0,0%,0%,0.2);
}

.sorting-wrapper {
  padding: 2rem;
  font-size: 1.5rem;
}

.sorting-wrapper h2 {
  font-weight: 500;
  margin-bottom: 1rem;
}

.news-card {
  display: -webkit-box;
  display: -ms-flexbox;
  /*display: flex;*/
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  margin: 0 auto;
  padding: 1rem;
  margin-bottom: 0.8rem;
  letter-spacing: 1px;
  background-color: white;
}

.news-card.show {
  -webkit-animation: show 300ms ease-in-out forwards;
  animation: show 300ms ease-in-out forwards;
}

.news-card.hide {
  -webkit-animation: hide 300ms ease-in-out forwards;
  animation: hide 300ms ease-in-out forwards;
  display: none;
}

.author-image {
  -ms-flex-preferred-size: 160px;
  flex-basis: 160px;
  -webkit-box-flex: 0.5;
  -ms-flex: 0.5;
  flex: 0.5;
  -ms-flex-item-align: center;
  -ms-grid-row-align: center;
  align-self: center;
}

.author-image img {
  display: block;
  margin: 0 auto;
  border-radius: 50%;
  -webkit-box-shadow: 0px 1px 5px #b3b3b3;
  box-shadow: 0px 1px 5px #b3b3b3;
}

.news-body {
  -webkit-box-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.news-info p {
  padding: .3rem 0;
}

.news-author {
  color: rgb(100, 100, 100);/*/#656565;*/
  font-weight: 400;
  padding: .5rem 0;
  /*text-shadow: 1px 1px 1px #b3b3b399;*/
}

.news-title {
  color: rgb(9, 9, 9);/*/#656565;*/
  font-weight: 410;
  padding: .5rem 0;
  text-decoration: none;
  word-wrap: break-word;
  /*text-shadow: 1px 1px 1px #b3b3b399;*/
}

.news-url {
  color: rgb(13, 102, 158);
  font-weight: 410;
  padding: .5rem 0;
  font-size: small;
  word-wrap: break-word;
}

.filter-wrapper {
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 3rem;
  margin: 2rem 0;
}

.radio-inputs>input {
  cursor: pointer;
}

.radio-inputs>label {
  cursor: pointer;
  font-size: 2rem;
  padding: 0 1.5rem 0 .5rem;
}

.search-input {
  position: relative;
  text-align: center;
}

.search-input input {
  padding: .7rem;
  font-size: 1.2rem;
  width: 0px;
  -webkit-transition: width 300ms ease-out;
  -o-transition: width 300ms ease-out;
  transition: width 300ms ease-out;
  border: 1px solid transparent;
  margin-left: 1.5rem;
}

.search-input input:focus {
  width: 250px !important;
}

.search-input img {
  cursor: pointer;
  position: absolute;
  width: 36px;
  top: 5px;
  right: 10px;
  vertical-align: middle;
}

@-webkit-keyframes show {
  from {
    opacity: 0;
    -webkit-transform: scale(0);
    transform: scale(0);
  }
  to {
    opacity: 1;
    -webkit-transform: scale(1);
    transform: scale(1);
  }
}

@keyframes show {
  from {
    opacity: 0;
    -webkit-transform: scale(0);
    transform: scale(0);
  }
  to {
    opacity: 1;
    -webkit-transform: scale(1);
    transform: scale(1);
  }
}

@-webkit-keyframes hide {
  from {
    opacity: 1;
    -webkit-transform: scale(1);
    transform: scale(1);
  }
  99% {
    opacity: 0;
    -webkit-transform: scale(0);
    transform: scale(0);
  }
  to {
    opacity: 0;
    -webkit-transform: scale(0);
    transform: scale(0);
  }
}

@keyframes hide {
  from {
    opacity: 1;
    -webkit-transform: scale(1);
    transform: scale(1);
  }
  99% {
    opacity: 0;
    -webkit-transform: scale(0);
    transform: scale(0);
  }
  to {
    opacity: 0;
    -webkit-transform: scale(0);
    transform: scale(0);
  }
}

@media (max-width: 760px) {
  body {
    font-size: 14px;
  }
  .code {
    top: 90%;
  }
  .filter-wrapper {
    padding: 1rem;
    -webkit-box-pack: center;
    -ms-flex-pack: center;
    justify-content: center;
    -webkit-box-orient: vertical;
    -webkit-box-direction: normal;
    -ms-flex-direction: column;
    flex-direction: column;
  }
  .filter-wrapper label {
    font-size: 1.3rem;
  }
  .search-input {
    margin: 1rem auto 0;
  }
  .search-input input {
    margin: 0;
  }
  .news-card {
    display: block;
  }
  .author-image {
    margin-bottom: 1rem;
  }
}



