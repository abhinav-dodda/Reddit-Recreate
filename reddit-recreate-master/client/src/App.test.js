import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

describe('App', () => {
  let component = null;

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    component = renderer.create(<App />);
    ReactDOM.unmountComponentAtNode(div);    
  });


  // fetchNews operation confirm
  it('fetch correctly', () => {    
      component.getInstance().fetchNews();
      expect(component.getInstance().state.after_news).toBe("");
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();    
  });

  // fetchPeriod operation confirm
  it('periodic fetch correctly', () => {    
      component.getInstance().fetchPeriod();
      expect(component.getInstance().state.news_count).toBe(0);
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();    
  });

  //element count after rendering
  it('NewsList element count', () => {    
      const wrapper = shallow(<App />);
      expect(wrapper.find(`NewsList`).length).toEqual(1);  
  });

});