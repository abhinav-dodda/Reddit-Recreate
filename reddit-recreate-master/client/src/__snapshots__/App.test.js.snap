// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App fetch correctly 1`] = `
<div
  className="App"
>
  <div
    className="header"
  >
    <img
      alt="logo"
      src="logo.png"
    />
  </div>
  <div
    className="container"
  >
    <div
      className="users-list"
    >
      <ul />
    </div>
  </div>
</div>
`;

exports[`App periodic fetch correctly 1`] = `
<div
  className="App"
>
  <div
    className="header"
  >
    <img
      alt="logo"
      src="logo.png"
    />
  </div>
  <div
    className="container"
  >
    <div
      className="users-list"
    >
      <ul />
    </div>
  </div>
</div>
`;
